src/system.rs

```diff
@@ -142,105 +156,485 @@ impl StageExecutor {
 
-// * implement QuerySet for tuples of queries
-// * likely actually wrapped in another struct, to cache the archetype sets for each query
-// * prepared queries will each re-use the archetype set results in their iterators so
-// that the archetype filters don't need to be run again - can also cache this between runs
-// and only append new archetype matches each frame
-// * per-query archetype matches stored as simple Vec<usize> - filter_archetypes() updates them and writes
-// the union of all queries into the BitSet provided, to be used to schedule the system as a whole
+    /// Iterates through all entity data that matches the query.
+    pub fn for_each<'a, 'data, T>(&'a mut self, mut f: T)
+    where
+        T: Fn(<<V as View<'data>>::Iter as Iterator>::Item),
+    {
+        self.iter().for_each(&mut f);
+    }
 
-trait QuerySet<'a>: Send + Sync {
-    type PreparedQueries: 'a;
+    /// Iterates through all entity data that matches the query in parallel.
+    #[cfg(feature = "par-iter")]
+    pub fn par_for_each<'a, T>(&'a mut self, f: T)
+    where
+        T: Fn(<<V as View<'a>>::Iter as Iterator>::Item) + Send + Sync,
+    {
+        self.par_iter_chunks().for_each(|mut chunk| {
+            for data in chunk.iter() {
+                f(data);
+            }
+        });
+    }
 
-    fn reads() -> Vec<ComponentTypeId>;
-    fn writes() -> Vec<ComponentTypeId>;
+    /// Gets a parallel iterator of chunks that match the query.
+    #[cfg(feature = "par-iter")]
+    pub fn par_iter_chunks<'a>(&'a mut self) -> impl ParallelIterator<Item = Chunk<'a, V>> {
+        self.iter_chunks().par_bridge()
+    }
+}
+
+pub trait QuerySet: Send + Sync {
+    type PreparedQueries;
     fn filter_archetypes(&mut self, world: &World, archetypes: &mut BitSet);
-    fn prepare(&'a self, world: &'a World) -> Self::PreparedQueries;
+    /// Safety: prepare call doesn't respect lifetimes of `self` and `world.
+    /// The returned value cannot outlive them.
+    unsafe fn prepare(&mut self, world: &World) -> Self::PreparedQueries;
+    // fn unprepare(prepared: Self::PreparedQueries) -> Self;
 }
 
-impl<'a> QuerySet<'a> for () {
-    type PreparedQueries = ();
+macro_rules! impl_queryset_tuple {
+    ($($ty: ident),*) => {
+        paste::item! {
+            #[allow(unused_parens, non_snake_case)]
+            impl<$([<$ty V>], [<$ty F>], )*> QuerySet for ($(Query<[<$ty V>], [<$ty F>]>, )*)
+            where
+                $([<$ty V>]: for<'v> View<'v>,)*
+                $([<$ty F>]: EntityFilter + Send + Sync,)*
+            {
+                type PreparedQueries = ( $(PreparedQuery<[<$ty V>], [<$ty F>]>, )*  );
+                fn filter_archetypes(&mut self, world: &World, bitset: &mut BitSet) {
+                    let ($($ty,)*) = self;
+
+                    $(
+                        let storage = world.storage();
+                        $ty.filter.iter_archetype_indexes(storage).for_each(|id| { bitset.insert(id); });
+                    )*
```

Comment:
Ultimately, I want the results of archetype filtering to be cached, and then re-used inside the prepared query so that we do not need to execute it a second time. I am not sure yet whether we can modify `Query` to support this, or if it will require a parallel implementation of query and its iterators (which is obviously undesirable). This does not need to be implemented for this PR, though.

Also, it would be ideal if we could resume `iter_archetype_indexes` from the index of the last archetype that was seen (in the previous frame), such that we only needed to inspect new archetypes each frame.